File: jas_ocaml/test/algorithms/align_fixture_test.ml
(** Consumes [test_fixtures/algorithms/align.json] entirely
    inside `dune runtest` and asserts the OCaml Align output
    matches the Rust and Swift reference for every vector. *)

open Yojson.Safe.Util
open Jas

(* Walk up from the current working directory until a [test_fixtures]
   sibling exists. [dune exec] runs from the project root while
   [dune runtest] runs from the test's build artefact dir
   ([_build/default/test/algorithms]) — the exact depth differs, so
   searching upward is the robust option. (A fixed relative path
   breaks under [dune runtest] because the fixtures live at the
   monorepo root, outside [_build].) *)
let fixtures_dir =
  let rec search dir depth =
    if depth > 8 then None
    else
      let candidate = Filename.concat dir "test_fixtures" in
      if Sys.file_exists candidate then Some candidate
      else
        let parent = Filename.dirname dir in
        if parent = dir then None
        else search parent (depth + 1)
  in
  match search (Sys.getcwd ()) 0 with
  | Some d -> d
  | None -> failwith "test_fixtures not found on path upward from cwd"

let fixture_path = Filename.concat fixtures_dir "algorithms/align.json"

let rect_of_json j =
  let to_f = function
    | `Int i -> float_of_int i
    | `Float f -> f
    | _ -> 0.0 in
  match to_list j with
  | [a; b; c; d] ->
    Element.make_rect (to_f a) (to_f b) (to_f c) (to_f d)
  | _ -> failwith "unexpected rect shape"

let bounds_of_json j =
  let to_f = function
    | `Int i -> float_of_int i
    | `Float f -> f
    | _ -> 0.0 in
  match to_list j with
  | [a; b; c; d] -> (to_f a, to_f b, to_f c, to_f d)
  | _ -> (0.0, 0.0, 0.0, 0.0)

let eps = 1e-4

let run_vector v =
  let op = to_string (member "op" v) in
  let rects = to_list (member "rects" v) |> List.map rect_of_json in
  let pairs = List.mapi (fun i e -> ([i], e)) rects in
  let use_preview = try to_bool (member "use_preview_bounds" v) with _ -> false in
  let bounds_fn =
    if use_preview then Align.preview_bounds
    else Align.geometric_bounds in
  let ref_kind = try to_string (member "kind" (member "reference" v))
                 with _ -> "selection" in
  let reference = match ref_kind with
    | "selection" -> Align.Selection (Align.union_bounds rects bounds_fn)
    | "artboard" ->
      Align.Artboard (bounds_of_json (member "bbox" (member "reference" v)))
    | "key_object" ->
      let idx = to_int (member "index" (member "reference" v)) in
      Align.Key_object { bbox = bounds_fn (List.nth rects idx); path = [idx] }
    | _ -> Align.Selection (0.0, 0.0, 0.0, 0.0) in
  let explicit_gap = match member "explicit_gap" v with
    | `Null -> None
    | `Float f -> Some f
    | `Int i -> Some (float_of_int i)
    | _ -> None in
  match op with
  | "align_left" -> Align.align_left pairs reference bounds_fn
  | "align_horizontal_center" -> Align.align_horizontal_center pairs reference bounds_fn
  | "align_right" -> Align.align_right pairs reference bounds_fn
  | "align_top" -> Align.align_top pairs reference bounds_fn
  | "align_vertical_center" -> Align.align_vertical_center pairs reference bounds_fn
  | "align_bottom" -> Align.align_bottom pairs reference bounds_fn
  | "distribute_left" -> Align.distribute_left pairs reference bounds_fn
  | "distribute_horizontal_center" -> Align.distribute_horizontal_center pairs reference bounds_fn
  | "distribute_right" -> Align.distribute_right pairs reference bounds_fn
  | "distribute_top" -> Align.distribute_top pairs reference bounds_fn
  | "distribute_vertical_center" -> Align.distribute_vertical_center pairs reference bounds_fn
  | "distribute_bottom" -> Align.distribute_bottom pairs reference bounds_fn
  | "distribute_vertical_spacing" ->
    Align.distribute_vertical_spacing pairs reference explicit_gap bounds_fn
  | "distribute_horizontal_spacing" ->
    Align.distribute_horizontal_spacing pairs reference explicit_gap bounds_fn
  | _ -> failwith ("unknown op: " ^ op)

let () =
  let json_str =
    let ic = open_in fixture_path in
    let n = in_channel_length ic in
    let s = really_input_string ic n in
    close_in ic; s in
  let fixture = Yojson.Safe.from_string json_str in
  let vectors = to_list (member "vectors" fixture) in
  let fixture_tests = List.map (fun v ->
    let name = to_string (member "name" v) in
    Alcotest.test_case name `Quick (fun () ->
      let actual = run_vector v in
      let expected = to_list (member "translations" v) in
      assert (List.length actual = List.length expected);
      List.iter2 (fun (a : Align.align_translation) e ->
        let e_path = to_list (member "path" e) |> List.map to_int in
        let e_dx = match member "dx" e with
          | `Float f -> f | `Int i -> float_of_int i | _ -> 0.0 in
        let e_dy = match member "dy" e with
          | `Float f -> f | `Int i -> float_of_int i | _ -> 0.0 in
        assert (a.path = e_path);
        assert (abs_float (a.dx -. e_dx) < eps);
        assert (abs_float (a.dy -. e_dy) < eps)
      ) actual expected)
  ) vectors in
  Alcotest.run "align_fixture" [ "vectors", fixture_tests ]

File: jas/algorithms/align_fixture_test.py
"""Consumes test_fixtures/algorithms/align.json entirely inside
`pytest` and asserts the Python Align output matches the Rust /
Swift / OCaml reference for every vector."""

import json
import os

from absl.testing import absltest

from algorithms import align as align_algo
from algorithms.align import (
    AlignReference, geometric_bounds, preview_bounds, union_bounds,
)
from geometry.element import Rect


_FIXTURE = os.path.join(
    os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))),
    "test_fixtures", "algorithms", "align.json",
)
_EPS = 1e-4


_OPS = {
    "align_left": align_algo.align_left,
    "align_horizontal_center": align_algo.align_horizontal_center,
    "align_right": align_algo.align_right,
    "align_top": align_algo.align_top,
    "align_vertical_center": align_algo.align_vertical_center,
    "align_bottom": align_algo.align_bottom,
    "distribute_left": align_algo.distribute_left,
    "distribute_horizontal_center": align_algo.distribute_horizontal_center,
    "distribute_right": align_algo.distribute_right,
    "distribute_top": align_algo.distribute_top,
    "distribute_vertical_center": align_algo.distribute_vertical_center,
    "distribute_bottom": align_algo.distribute_bottom,
}
_SPACING_OPS = {
    "distribute_vertical_spacing": align_algo.distribute_vertical_spacing,
    "distribute_horizontal_spacing": align_algo.distribute_horizontal_spacing,
}


def _run_vector(v):
    rects = [Rect(x=r[0], y=r[1], width=r[2], height=r[3]) for r in v["rects"]]
    pairs = [((i,), r) for i, r in enumerate(rects)]
    use_preview = v.get("use_preview_bounds", False)
    bounds_fn = preview_bounds if use_preview else geometric_bounds

    ref = v.get("reference") or {}
    kind = ref.get("kind", "selection")
    if kind == "artboard":
        bbox = ref.get("bbox", [0, 0, 0, 0])
        reference = AlignReference.artboard(tuple(bbox))
    elif kind == "key_object":
        idx = ref["index"]
        reference = AlignReference.key_object(bounds_fn(rects[idx]), (idx,))
    else:
        reference = AlignReference.selection(union_bounds(rects, bounds_fn))

    explicit_gap = v.get("explicit_gap")
    op = v["op"]
    if op in _OPS:
        return _OPS[op](pairs, reference, bounds_fn)
    if op in _SPACING_OPS:
        return _SPACING_OPS[op](pairs, reference, explicit_gap, bounds_fn)
    raise ValueError(f"unknown op: {op}")


class AlignFixtureTest(absltest.TestCase):
    pass


def _load_vectors():
    with open(_FIXTURE) as f:
        return json.load(f)["vectors"]


def _make_test(vector):
    def test(self):
        actual = _run_vector(vector)
        expected = vector["translations"]
        self.assertEqual(len(actual), len(expected),
                         f"translation count mismatch in {vector['name']}")
        for a, e in zip(actual, expected):
            self.assertEqual(list(a.path), e["path"],
                             f"path mismatch in {vector['name']}")
            self.assertAlmostEqual(a.dx, e["dx"], delta=_EPS,
                                   msg=f"dx mismatch in {vector['name']}")
            self.assertAlmostEqual(a.dy, e["dy"], delta=_EPS,
                                   msg=f"dy mismatch in {vector['name']}")
    return test


for _vec in _load_vectors():
    _name = "test_" + _vec["name"]
    setattr(AlignFixtureTest, _name, _make_test(_vec))


if __name__ == "__main__":
    absltest.main()
